### PR TITLE
Remove built cache of previous git commits.

### PR DESCRIPTION
**Pull request by spoojaryTT**
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/14619)

### Problem description
Previously built files might still be reused when the tt_metal is updated.

### What's changed
1. During build, in cmake, add compile defines for GIT_COMMIT_HASH. Use the `VERSION_HASH` set by parent CMakeLists.txt
2. Build.cpp would check for the git hash define (GIT_COMMIT_HASH), if hash is not found, proceed without changing any previously built folders and root folder will be 'built/'
3. If git hash is found, the root folder will be 'built/GIT_COMMIT_HASH'.
4. If git hash is found, and TT_METAL_SKIP_DELETING_BUILT_CACHE is NOT set, delete all directories in 'built' except for 'built/GIT_COMMIT_HASH' folder
5. If git hash is found, and TT_METAL_SKIP_DELETING_BUILT_CACHE is set, no directory is deleted. Root for current build will be 'built/GIT_COMMIT_HASH'

### Checklist
- [x] Post commit CI passes - https://github.com/tenstorrent/tt-metal/actions/runs/11966119330
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tt_metal/jit_build/CMakeLists.txt
Comment (blozano-tt) [inline tt_metal/jit_build/CMakeLists.txt:22]: What happens if:

A) We are building from a tar.gz source blob and there is no .git directory? Will things still function?

B) Git is not installed?

I think we need to cover these two cases before this is merged.

**Review by blozano-tt — CHANGES_REQUESTED** [review]
Looks good overall. I have only one gripe.

File: tt_metal/jit_build/CMakeLists.txt
Comment (tt-rkim) [reply tt_metal/jit_build/CMakeLists.txt:22]: Yes, don't we have a general rule of not relying on source control to perform a build.

Looping in @afuller-TT as well

Comment (spoojaryTT) [reply tt_metal/jit_build/CMakeLists.txt:22]: if the command returns any error, the "GIT_COMMIT_HASH" will not be defined, and in that case we will create directories like how we do now, without any hash (built/.../), instead of "built/GIT_COMMIT_HASH/... "
I will test out the two cases you mentioned as well.

Comment (tt-rkim) [reply tt_metal/jit_build/CMakeLists.txt:22]: Since it's a build parameter, how can we test this in a gtest?

Comment (blozano-tt) [inline tt_metal/jit_build/CMakeLists.txt:22]:
```suggestion

find_program(GIT_EXECUTABLE git)

if(GIT_EXECUTABLE)
    execute_process(
        COMMAND
            git rev-parse --short=10 HEAD
        RESULT_VARIABLE git_parse_result
        OUTPUT_VARIABLE git_commit_hash_value
        OUTPUT_STRIP_TRAILING_WHITESPACE
    )

    if(git_parse_result EQUAL "0")
        target_compile_definitions(jit_build PUBLIC "-DGIT_COMMIT_HASH=\"${git_commit_hash_value}\"")
    endif()
endif()
```

File: tt_metal/jit_build/CMakeLists.txt
Comment (afuller-TT) [inline tt_metal/jit_build/CMakeLists.txt:12]: Did this work as-is?  🤔 

```suggestion
find_program(GIT_EXECUTABLE git)
```

Comment (spoojaryTT) [reply tt_metal/jit_build/CMakeLists.txt:12]: My bad, I was testing what happens if 'git' is not present in the system. Ended up committing that change.
It turns out, find_program will find 'git', even if I ask it find 'gi'.
I feel the 'result_variable' is more reliable to catch any errors, find_program doesn't catch all issues. If a system doesn't have 'git', but has some other application whose name starts with 'git', it might return that application.
@blozano-tt

**Review by afuller-TT — APPROVED** [review]
I'm good with this.  The CMake changes look good (once we fix the name of `git`).

I'm not familiar with the JIT build and haven't combed through the code to know why we're loading stale items.  This approach feels like a hammer to me, but again I don't know the nuances so you know what's needed here.

File: tt_metal/jit_build/CMakeLists.txt
Comment (afuller-TT) [reply tt_metal/jit_build/CMakeLists.txt:12]: All `find*` will store positive results in the CMakeCache.txt for efficiency the next time.  Unfound things will be re-evaluated in case they've shown up since the last run.

That means to test a found -> not found, you'll need to remove the entry from CMakeCache.txt (or just blow away the entire file).

**Review by blozano-tt — APPROVED** [review]

**Comment by blozano-tt** [comment]
> I'm good with this. The CMake changes look good (once we fix the name of `git`).
> 
> I'm not familiar with the JIT build and haven't combed through the code to know why we're loading stale items. This approach feels like a hammer to me, but again I don't know the nuances so you know what's needed here.

I also find it weird.

I asked previously why we can't just put things in `/tmp/` and let them get deleted after the process completes.

**Comment by tt-rkim** [comment]
We can't really test this change in CI without doing another build since it's a compile define

**Comment by spoojaryTT** [comment]
> We can't really test this change in CI without doing another build since it's a compile define

**Comment by spoojaryTT** [comment]
> We can't really test this change in CI without doing another build since it's a compile define

If we want to avoid compile define, we can move finding of git hash to inside metal code. Use system() and pipe the output to a file.  With this, we can probably test that root path is as expected.
Thoughts @tt-rkim @pgkeller

**Comment by tt-rkim** [comment]
Please do not add runtime code to get source control information. That'll break pipelines, anyway.

I'm more concerned of how you're gonna write an automated test for this at all.

Also, once @afuller-TT 's changes from here https://github.com/tenstorrent/tt-metal/pull/15400/files go in, we can change your CMake changes to parse data from `VERSION_HASH` so it's all in one place.

Can we address @blozano-tt and @afuller-TT 's comments about this being a big hammer? I know we've thought about this before. Are there no smaller hammers we can use to invalidate the cache of previous commits?

**Comment by pgkeller** [comment]
> Please do not add runtime code to get source control information. That'll break pipelines, anyway.
> 
> I'm more concerned of how you're gonna write an automated test for this at all.
> 
> Also, once @afuller-TT 's changes from here https://github.com/tenstorrent/tt-metal/pull/15400/files go in, we can change your CMake changes to parse data from `VERSION_HASH` so it's all in one place.
> 
> Can we address @blozano-tt and @afuller-TT 's comments about this being a big hammer? I know we've thought about this before. Are there no smaller hammers we can use to invalidate the cache of previous commits?

discussion prior to the PR (on devlopers channel, I think) people were worried about leaving stale files behind, hence the delete.  the way I see it:
1) We need a key in the path.  how we get the key I don't care, sounds like we need to wait for @afuller-TT's change
2) Optionally we can clean up.  Sounds like developers want that and you don't.  We can just delete that

does this work?

**Comment by tt-rkim** [comment]
We can do 2 for now until @afuller-TT 's change hits, then we should be able to do 1 and easily provide what the hash here while accounting for some release (git archive) developer flows.

How's that sound, or am I misunderstanding? @afuller-TT @pgk

**Review by tt-rkim — APPROVED** [review]
Build looks good, but I still don't see a test to test out the other build path where `VERSION_HASH` is not defined? Am I blind?

File: tt_metal/jit_build/CMakeLists.txt
Comment (afuller-TT) [inline tt_metal/jit_build/CMakeLists.txt:13]: `PUBLIC`?  This will have serious implications on builds.  It'll force a new define on all consumers.  I don't see anyone outside of `jit_build` using it, so that's just unnecessary make-work for the compiler.

```suggestion
    target_compile_definitions(jit_build PRIVATE GIT_COMMIT_HASH=${VERSION_HASH})
```

**Review by pgkeller — APPROVED** [review]

File: tt_metal/jit_build/CMakeLists.txt
Comment (tt-rkim) [reply tt_metal/jit_build/CMakeLists.txt:14]: We do, but I'll leave that up to you guys.

**Comment by tt-rkim** [comment]
Sure, then we can leave it as a dev feature with no testing.

Approved~!

**Pull request merged**
